param: changes for making param set/get work with ArduPilot

ykhedar opened the pull request:
This pull request contains the minimal necessary changes needed in the MAVSDK to make param set/get work with Ardupilot. Following the discussion from https://github.com/mavlink/MAVSDK/issues/1683, I decided to try the suggestion B from @julianoes 

> B) Always download all params on startup to then have the type cached when the function is used. This would take longer but make the API easier. What do you think?

I had to, however create a different function to deal with the param set for APM

```bash
float MAVLinkParameters::ParamValue::get_4_float_bytes() const
float MAVLinkParameters::ParamValue::get_4_float_bytes_apm() const
```
As mentioned before, Ardupilot likes to have direct cast to float instead of the re-interpret cast used in MAVSDK for PX4 param set. More details can be found in https://mavlink.io/en/services/parameter.html#ardupilot

Awaiting valued feedback on this.

JonasVautherin commented on src/mavsdk/core/mavlink_parameters.cpp:
```diff
@@ -311,7 +311,19 @@ void MAVLinkParameters::do_work()
             } else {
                 // Param set is intended for Autopilot only.
-                mavlink_msg_param_set_pack(
+                if (_parent.autopilot() == SystemImpl::Autopilot::ArduPilot) {
```
Maybe a nitpick, but what about having the `if` just set a temporary variable like this?

```c++
float my_param;
if (_parent.autopilot() == SystemImpl::Autopilot::ArduPilot) {
    my_param = work->param_value.get_4_float_bytes_apm();
} else {
    my_param = work->param_value.get_4_float_bytes();
}
```

This makes it clear that the difference between APM and PX4 is only for this param `my_param` (well, it needs a better name), and all the other parameters given to `mavlink_msg_param_set_pack` are the same.

What do you think?

JonasVautherin commented on src/mavsdk/core/system_impl.cpp:
```diff
@@ -767,15 +791,29 @@ std::pair<MAVLinkParameters::Result, int> SystemImpl::get_param_int(const std::s
     MAVLinkParameters::ParamValue value_type;
-    value_type.set<int32_t>(0);
-
+    //value_type.set<int32_t>(0);
```
```suggestion

```

ykhedar replied:
Done.

ykhedar commented on src/mavsdk/core/system_impl.cpp:
```diff
@@ -643,14 +646,35 @@ MAVLinkParameters::Result SystemImpl::set_param_float(const std::string& name, f
 std::map<std::string, MAVLinkParameters::ParamValue> SystemImpl::get_all_params()
 {
-    return _params.get_all_params();
+    _param_map = _params.get_all_params();
+    // for (auto const& x : _param_map)
+    // {
+    //     std::cout << x.first  // ParamValue ID
+    //             << ':' 
+    //             << x.second // ParamValue Value 
+    //             << ':' 
+    //             << x.second.get_mav_param_type() // ParamValue Type
+    //             << std::endl;
+    // }
+    return _param_map;
```
Sadly yes, the problem is we need to keep a cache of the parameter and I could only think of this way right now. So as soon as the System is created and it inits the plugins e.g. param plugin, the param plugin in its init method sends a request to get_all_params() which then populates the _param_map in System class and then also returns it to the param plugin.
let me know if you have a better idea ;).

JonasVautherin commented on src/mavsdk/core/system_impl.cpp:
```diff
@@ -643,14 +644,25 @@ MAVLinkParameters::Result SystemImpl::set_param_float(const std::string& name, f
 {
     MAVLinkParameters::ParamValue param_value;
-    param_value.set<int32_t>(value);
+    //param_value.set<int32_t>(value);
```
```suggestion

```

JonasVautherin commented:
The CI is not happy, and you should fix the style with something like `./tools/fix_style.sh .` :+1: :blush:

julianoes commented:
@ykhedar I've rebased your PR, fixed style issues, and force pushed it to your branch.

I also added a test to check `get_all_params`:
```
cmake --build build/default/ -j6 && build/default/src/integration_tests/integration_tests_runner --gtest_filter="SitlTest.GetAllParams"
```
(For ArduPilot, don't forget to change `add_udp_connection()` to `add_udp_connection(14550)`.

The test fails for me for me for ArduPilot,
```
[05:29:00|Info ] Int params: 0 (param.cpp:149)
[05:29:00|Info ] Float params: 0 (param.cpp:155)
[05:29:00|Info ] Combined params: 0 (param.cpp:161)
```

and PX4:
```
[05:30:13|Error] Error: unknown mavlink param type:  (mavlink_parameters.cpp:965)
[05:30:13|Error] Error: unknown mavlink param type:  (mavlink_parameters.cpp:965)
[05:30:14|Info ] Int params: 0 (param.cpp:149)
[05:30:14|Info ] Float params: 0 (param.cpp:155)
[05:30:14|Info ] Combined params: 0 (param.cpp:161)
```

so something is not quite right yet.
It works for PX4 before that.

One thing that I think of changing is to only get the param that you want to set. That way we don't have to get all params which is not necessarily something that everyone wants to do, always.

julianoes commented:
FYI @ykhedar I'm currently reworking this.

ykhedar commented:
> @ykhedar I've rebased your PR, fixed style issues, and force pushed it to your branch.
> 
> I also added a test to check `get_all_params`:
> 
> ```
> cmake --build build/default/ -j6 && build/default/src/integration_tests/integration_tests_runner --gtest_filter="SitlTest.GetAllParams"
> ```
> 
> (For ArduPilot, don't forget to change `add_udp_connection()` to `add_udp_connection(14550)`.
> 
> The test fails for me for me for ArduPilot,
> 
> ```
> [05:29:00|Info ] Int params: 0 (param.cpp:149)
> [05:29:00|Info ] Float params: 0 (param.cpp:155)
> [05:29:00|Info ] Combined params: 0 (param.cpp:161)
> ```
> 
> and PX4:
> 
> ```
> [05:30:13|Error] Error: unknown mavlink param type:  (mavlink_parameters.cpp:965)
> [05:30:13|Error] Error: unknown mavlink param type:  (mavlink_parameters.cpp:965)
> [05:30:14|Info ] Int params: 0 (param.cpp:149)
> [05:30:14|Info ] Float params: 0 (param.cpp:155)
> [05:30:14|Info ] Combined params: 0 (param.cpp:161)
> ```
> 
> so something is not quite right yet. It works for PX4 before that.
> 
> One thing that I think of changing is to only get the param that you want to set. That way we don't have to get all params which is not necessarily something that everyone wants to do, always.

Hi @julianoes Thanks for taking a look at the PR. I am not able to work on it for at least a week more due to the other project so please go ahead and rework if you have the time. 

Last time I checked the tests not passing for PX4 set param was most probably because the params were not being cached at the right time. I was looking into it but then dragged into the other project.

I believe caching all params on first connection is not a bad idea. I have seen this behaviour in Mavproxy and really like that, as soon as it is connected to the vehicle it saves all params to a a file. Since MAVSDK is usually connected over a wire to the Autopilot, should not be a huge problem speed or lost packet wise. What do you think?

ykhedar commented:
@julianoes There was a small bug(wrong function name) which was causing issues for CI. I fixed it in my last commit and CI seems to be happy now. Are you still working on something for this pull request or we can rebase and merge it into the main branch?

julianoes commented:
@ykhedar I'm having a look now, sorry for the delay.

julianoes commented:
@ykhedar thanks for fixing that, I was wondering what was going on!

julianoes reviewed (approved):
Reviewed again, let's get this in and start using it.
